src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.cs

```diff
@@ -0,0 +1,308 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.CodeStyle2
+{
+    /// <summary>
+    /// This is the core class a code-style feature needs to derive from.  All logic related to the
+    /// feature will then be contained in this class.  This class will take care of many bit of
+    /// common logic that all code style providers would have to care about and can thus do that
+    /// logic in a consistent fashion without all providers having to do the same.  For example,
+    /// this class will check the current value of the code style option. If it is 'refactoring
+    /// only', it will not bother running any of the DiagnosticAnalyzer codepaths, and will only run
+    /// the CodeRefactoringProvider codepaths.
+    /// </summary>
+    internal abstract class AbstractCodeStyle2Provider<
+        TOptionKind, 
+        TSyntaxKind,
+        TCodeStyleProvider>
+        where TSyntaxKind : struct
+        where TCodeStyleProvider : AbstractCodeStyle2Provider<TOptionKind, TSyntaxKind, TCodeStyleProvider>, new()
+    {
+        private readonly Option<CodeStyleOption<TOptionKind>> _option;
+        private readonly string _descriptorId;
+        private readonly LocalizableString _title;
+        private readonly LocalizableString _message;
+        private CodeStyle2CodeFixProvider _fixProvider;
+
+        protected AbstractCodeStyle2Provider(
+            Option<CodeStyleOption<TOptionKind>> option,
+            string descriptorId,
+            LocalizableString title,
+            LocalizableString message)
+        {
+            _option = option;
+            _descriptorId = descriptorId;
+            _title = title;
+            _message = message;
+        }
+
+        #region diagnostic analyzer
+
+        protected abstract DiagnosticAnalyzerCategory GetDiagnosticAnalyzerCategory();
+        protected abstract bool DiagnosticsForOpenFileOnly(Workspace workspace);
+
+        protected virtual Action<CodeBlockAnalysisContext> GetCodeBlockAction() => null;
+        protected virtual (ImmutableArray<TSyntaxKind>, Action<SyntaxNodeAnalysisContext>) GetSyntaxNodeAction() => default;
+
+        private void DiagnosticAnalyzerInitialize(AnalysisContext context)
+        {
+            var analyzeCodeBlock = GetCodeBlockAction();
+            if (analyzeCodeBlock != null)
+            {
+                context.RegisterCodeBlockAction(c => AnalyzeIfEnabled(c, analyzeCodeBlock));
```

Comment:
critically, we want to consolidate the logic about checking if the analyzer should run at all.  i.e. if the user has their option set to 'none' or 'refactoring only' then we do not want the analyzer to run at all.  

To that end, we don't let hte subclass have direct access to the AnalysisContext . Instead, we query them to see what they would like to register.  And then we do the registration for them.  *however*, before we call into their callback, we first do a check if analysis should happen at all.  If not, we never even bother calling into the callback.